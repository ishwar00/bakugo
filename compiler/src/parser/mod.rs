use miette::NamedSource;
use pest::{
    error::Error,
    iterators::{Pair, Pairs},
    Parser,
};

use crate::ast::{BakugoParsingErrorDisplay, Node, SourceFile};

#[derive(pest_derive::Parser)]
#[grammar = "bakugo.pest"]
pub struct BakugoParser;

pub fn parse_string(s: &str) -> Result<Pairs<'_, Rule>, Box<Error<Rule>>> {
    match BakugoParser::parse(Rule::SourceFile, s) {
        Ok(ast) => Ok(ast),
        Err(err) => Err(Box::new(err)),
    }
}

pub fn construct_ast(
    source: NamedSource,
    pair: Pair<'_, Rule>,
) -> Result<SourceFile<'_>, BakugoParsingErrorDisplay> {
    match SourceFile::parse(pair) {
        Ok(s) => Ok(s),
        Err(bakugo_err) => Err(BakugoParsingErrorDisplay::from_error(bakugo_err, source)),
    }
}
