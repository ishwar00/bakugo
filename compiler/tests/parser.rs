use std::fs;

use bakugo::parser::parse_string;
use insta::{assert_snapshot, assert_yaml_snapshot, glob};

#[test]
fn test_parser() {
    glob!("examples/*.bakugo", |path| {
        let input = fs::read_to_string(path).unwrap();
        let parsed = parse_string(&input);
        match parsed {
            Ok(parsed) => assert_yaml_snapshot!(parsed),
            Err(err) => assert_snapshot!(err.to_string()),
        }
    });
}
