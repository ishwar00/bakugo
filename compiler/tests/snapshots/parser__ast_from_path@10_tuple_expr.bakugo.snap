---
source: tests/parser.rs
expression: parsed
input_file: tests/examples/10_tuple_expr.bakugo
---
SourceFile {
    top_level: [
        FnDecl(
            FnDecl {
                name: Ident {
                    value: "main",
                    span: Span {
                        str: "main",
                        start: 5,
                        end: 9,
                    },
                },
                params: [],
                result_kind: Some(
                    Tuple {
                        kinds: [
                            Simple {
                                name: "int",
                                span: Span {
                                    str: "int",
                                    start: 13,
                                    end: 16,
                                },
                            },
                            Simple {
                                name: "int",
                                span: Span {
                                    str: "int",
                                    start: 18,
                                    end: 21,
                                },
                            },
                        ],
                        span: Span {
                            str: "int, int",
                            start: 13,
                            end: 21,
                        },
                    },
                ),
                body: StatementList(
                    [
                        Declaration(
                            Var(
                                VarConstDecl {
                                    name: Ident {
                                        value: "a",
                                        span: Span {
                                            str: "a",
                                            start: 34,
                                            end: 35,
                                        },
                                    },
                                    kind: None,
                                    expr: TupleExpr(
                                        [
                                            Integer {
                                                value: 34,
                                                span: Span {
                                                    str: "34",
                                                    start: 39,
                                                    end: 41,
                                                },
                                            },
                                            TupleExpr(
                                                [
                                                    Integer {
                                                        value: 4,
                                                        span: Span {
                                                            str: "4",
                                                            start: 44,
                                                            end: 45,
                                                        },
                                                    },
                                                    Integer {
                                                        value: 45,
                                                        span: Span {
                                                            str: "45",
                                                            start: 47,
                                                            end: 49,
                                                        },
                                                    },
                                                ],
                                            ),
                                        ],
                                    ),
                                },
                            ),
                        ),
                        Declaration(
                            Var(
                                VarConstDecl {
                                    name: Ident {
                                        value: "a",
                                        span: Span {
                                            str: "a",
                                            start: 55,
                                            end: 56,
                                        },
                                    },
                                    kind: None,
                                    expr: TupleExpr(
                                        [
                                            Integer {
                                                value: 34,
                                                span: Span {
                                                    str: "34",
                                                    start: 60,
                                                    end: 62,
                                                },
                                            },
                                            TupleExpr(
                                                [
                                                    Integer {
                                                        value: 4,
                                                        span: Span {
                                                            str: "4",
                                                            start: 65,
                                                            end: 66,
                                                        },
                                                    },
                                                    Integer {
                                                        value: 45,
                                                        span: Span {
                                                            str: "45",
                                                            start: 68,
                                                            end: 70,
                                                        },
                                                    },
                                                ],
                                            ),
                                        ],
                                    ),
                                },
                            ),
                        ),
                        Declaration(
                            Var(
                                VarConstDecl {
                                    name: Ident {
                                        value: "b",
                                        span: Span {
                                            str: "b",
                                            start: 77,
                                            end: 78,
                                        },
                                    },
                                    kind: None,
                                    expr: TupleExpr(
                                        [
                                            Integer {
                                                value: 34,
                                                span: Span {
                                                    str: "34",
                                                    start: 82,
                                                    end: 84,
                                                },
                                            },
                                        ],
                                    ),
                                },
                            ),
                        ),
                    ],
                ),
                span: Span {
                    str: "func main() (int, int) {\n\tvar (\n\t\ta = (34, (4, 45));\n\t\ta = (34, (4, 45),);\n\t\tb = (34,);\n\t);\n}",
                    start: 0,
                    end: 93,
                },
            },
        ),
    ],
}
